app/assets/javascripts/oxalis/model/reducers/skeletontracing_reducer_helpers.js

```diff
@@ -117,156 +117,215 @@ export function deleteNode(
+  return getSkeletonTracing(state.tracing).chain(skeletonTracing => {
+    const { allowUpdate } = skeletonTracing.restrictions;
+
+    if (allowUpdate) {
+      if (sourceTree.treeId !== targetTree.treeId) {
+        // The two selected nodes are in different trees
+        console.error(
+          "Tried two delete an edge that was not there, the two nodes are in different trees.",
+        );
+        return Maybe.Nothing();
+      }
+
+      console.log(sourceTree.edges, sourceNode.id, targetNode.id);
+      const deletedEdge = sourceTree.edges.find(
+        edge =>
+          (edge.source === sourceNode.id && edge.target === targetNode.id) ||
+          (edge.source === targetNode.id && edge.target === sourceNode.id),
+      );
+
+      if (deletedEdge == null) {
+        // The two selected nodes do not share an edge
+        console.error("Tried two delete an edge that was not there.");
+        return Maybe.Nothing();
+      }
+
+      return Maybe.Just(
+        splitTreeByNodes(
+          state,
+          skeletonTracing,
+          sourceTree,
+          [sourceNode.id, targetNode.id],
+          [deletedEdge],
+          timestamp,
+        ),
+      );
+    } else {
+      return Maybe.Nothing();
+    }
+  });
+}
+
+function splitTreeByNodes(
+  state: OxalisState,
+  skeletonTracing: SkeletonTracingType,
+  activeTree: TreeType,
+  neighborIds: Array<number>,
+  deletedEdges: Array<EdgeType>,
+  timestamp: number,
+): TreeMapType {
+  // Use split-algorithmus. If we delete a node which is only connected via one edge,
+  // this algorithmus will only produce one tree (which reuses the old one)
+
+  let newTrees = skeletonTracing.trees;
+
+  // Build a hashmap which contains for each node all edges leading/leaving into/from the node
+  const nodeToEdgesMap: { [number]: Array<EdgeType> } = {};
+  activeTree.edges.forEach(edge => {
+    if (nodeToEdgesMap[edge.source]) {
+      nodeToEdgesMap[edge.source].push(edge);
+    } else {
+      nodeToEdgesMap[edge.source] = [edge];
+    }
+    if (nodeToEdgesMap[edge.target]) {
+      nodeToEdgesMap[edge.target].push(edge);
+    } else {
+      nodeToEdgesMap[edge.target] = [edge];
+    }
+  });
+
+  // Traverse from active node in all directions (i.e., use each edge) and
+  // remember which edges were already visited
+  const visitedEdges = {};
+  const getEdgeHash = edge => `${edge.source}-${edge.target}`;
+
+  // Mark edges of deleted node as visited
+  deletedEdges.forEach(deletedEdge => {
+    visitedEdges[getEdgeHash(deletedEdge)] = true;
+  });
+
+  const traverseTree = (nodeId: number, newTree: TemporaryMutableTreeType) => {
+    const edges = nodeToEdgesMap[nodeId];
+
+    newTree.nodes[nodeId] = activeTree.nodes[nodeId];
+
+    for (const edge of edges) {
+      const edgeHash = getEdgeHash(edge);
+      if (visitedEdges[edgeHash]) {
+        continue;
+      }
+      visitedEdges[edgeHash] = true;
+      newTree.edges.push(edge);
+
+      traverseTree(edge.source, newTree);
```

Comment:
Be careful when merging the master, since we changed the splitting to be iterate.

Reply:
I think this was done on the ...perf branch, not yet on master :)